### PR TITLE
fix(bash): add case for vscode integrated terminal

### LINKED ISSUE #5816: `starship init bash` in vscode: integrated terminal crashes due to infinite loop
pkg_version:1.17.1
branch:master
commit_hash:
build_time:2024-01-02 18:32:30 +00:00
build_env:rustc 1.75.0 (82e1608df 2023-12-21),

### Current Behavior

When using `eval "$(starship init bash)"` in my bashrc, when I start Visual Studio Code integrated terminal, it loads correctly, but if I try to source my bashrc manually during the session, the terminal crashes (and closes).

### Expected Behavior

The vscode integrated terminal does not crash when running `eval "$(starship init bash)"` twice in the same terminal.

### More context

- Tested on Windows 10 with vscode 1.87.0, with lot of extensions and WSL as the terminal backend
- Tested on Ubuntu 22.04 with vscode 1.86.2, fresh install, no extensions

#### My debug results

**TL;DR**: infinite loop caused by starship preserving `PROMPT_COMMAND`, and by vscode also preserving `PROMPT_COMMAND`.

---

<details>

**<summary>Debugging details (click to unfold)</summary>**

In vscode, with shell integration enabled (NOT the experimental ones), `PROMPT_COMMAND` is set to `__vsc_prompt_cmd_original` (after sourcing `bashrc`), and it overrides the value set by starship.

```bash
$ # This is a fresh terminal in vscode, with shell integration enabled, and with starship init in bashrc.

$ type -t starship_precmd
function

$ echo $PROMPT_COMMAND 
__vsc_prompt_cmd_original

$ type __vsc_prompt_cmd_original
__vsc_prompt_cmd_original is a function
__vsc_prompt_cmd_original () 
{ 
    __vsc_status="$?";
    __vsc_restore_exit_code "${__vsc_status}";
    for cmd in "${__vsc_original_prompt_command[@]}";
    do
        eval "${cmd:-}";
    done;
    __vsc_precmd
}

$ echo "${__vsc_original_prompt_command[@]}"
starship_precmd

$ type __vsc_precmd
__vsc_precmd is a function
__vsc_precmd () 
{ 
    __vsc_command_complete "$__vsc_status";
    __vsc_current_command="";
    __vsc_update_prompt
}
```

vscode preserves `PROMPT_COMMAND` and executes it before its own `__vsc_precmd` function.

But starship does the same: it preserves `PROMPT_COMMAND` and executes it before its own logic in `starship_precmd`.

We can see it if we try to `echo $_PRESERVED_PROMPT_COMMAND` in `.bashrc`. Because this is prompt related stuff, the `bashrc` runs to completion, and the infinite loop is triggered after that, when the prompt needs to be printed.

```bash
_PRESERVED_PROMPT_COMMAND:
• me@host:~ fancyprompt
$ # this was a new terminal opened in vscode

• me@host:~ fancyprompt
$ . ~/.bashrc

_PRESERVED_PROMPT_COMMAND: __vsc_prompt_cmd_original
# (then terminal crashes)
```

**This is the source of the infinite loop.**

- At terminal startup, `__vsc_prompt_cmd_original` does not exist, and `starship_precmd` executes just fine
- When evaluating `starship init bash` again:
  - `PROMPT_COMMAND` is set to `__vsc_prompt_cmd_original` (and it will execute `starship_precmd`)
  - starship saves this `PROMPT_COMMAND` as `_PRESERVED_PROMPT_COMMAND`
  - `__vsc_prompt_cmd_original` executes `starship_precmd`
  -  `starship_precmd` executes `__vsc_prompt_cmd_original` as `_PRESERVED_PROMPT_COMMAND`
  - `__vsc_prompt_cmd_original` executes `starship_precmd`
  - infinite loop

</details>

---

### Possible Solution

In [starship.bash](https://github.com/starship/starship/blob/master/src/init/starship.bash), add a verification specific to vscode: if `${__vsc_original_prompt_command[@]}` contains `starship_precmd`, do not redefine `PROMPT_COMMAND`.

### Environment

- Starship version: 1.17.1
- bash version: GNU bash, version 5.2.26(1)-release (x86_64-unknown-linux-gnu)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
- Operating system: Linux Unknown
- Terminal emulator: vscode 1.86.2
- Git Commit Hash: 
- Branch/Tag: master
- Rust Version: rustc 1.75.0 (82e1608df 2023-12-21)
- Rust channel:  release
- Build Time: 2024-01-02 18:32:30 +00:00

---

### Relevant Shell Configuration

<details>

**<summary>`bashrc`</summary>**

```bash
# minimalist ~/.bashrc for testing

# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac

shopt -s histappend
shopt -s checkwinsize
HISTCONTROL=ignoreboth
HISTSIZE=1000
HISTFILESIZE=2000

# Add ~/.local/bin in PATH if not already present
case :$PATH:
  in *:$HOME/.local/bin:*) ;; # do nothing, it's there
     *) export PATH="$HOME/.local/bin:$PATH" ;;
esac

# Fallback PS1
PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\$ '

command -v starship &>/dev/null && eval "$(starship init bash)"

alias ll='ls -alF'
```

</details>

### Starship Configuration

Default (no file)

### PR DESCRIPTION
**Pull request by NeodymiumFerBore**
#### Description

In starship.bash init script, handle vscode integrated terminal `PROMPT_COMMAND` override.

#### Motivation and Context

Fix an issue where manually sourcing a file evaluating `starship init bash` (typically a bashrc) would crash vscode integrated terminal if starship init had already been evaluated at terminal startup.

Closes #5816 

#### How Has This Been Tested?

- Tested on Windows 10 with vscode 1.87.0, with lot of extensions and WSL as the terminal backend
- Tested on Ubuntu 22.04 with vscode 1.86.2, fresh install, no extensions

<details>

<summary>tests details</summary>

After reproducing the bug described in issue #5816 :

##### In vscode

- open a new bash in vscode terminal
- assess that starship init was evaluated with the fancy prompt appearing
- `cd` into starship git repository
- `git checkout fix/vscode-infinite-loop`
- `::STARSHIP::() { starship "$@"; }`
- `eval "$(cat ./src/init/starship.bash)"`
- the terminal does not crash
- starship environment variables are correctly setup, the prompt works as expected
- `__vsc_original_prompt_command=()`
- `eval "$(cat ./src/init/starship.bash)"`
- the terminal does not crash, `${__vsc_original_prompt_command[@]}` contains `starship_precmd` as expected

##### Outside vscode

- tested in WSL and Ubuntu 22.04
- open a new terminal
- assess that starship init was evaluated with the fancy prompt appearing
- `cd` into starship git repository
- `git checkout fix/vscode-infinite-loop`
- `::STARSHIP::() { starship "$@"; }`
- `eval "$(cat ./src/init/starship.bash)"`
- the terminal does not crash

##### Both inside and outside vscode: check we didn't break something

- `unset PROMPT_COMMAND`
- the PS1 does not change anymore, we "broke" starship prompt
- `eval "$(cat ./src/init/starship.bash)"`
- `PROMPT_COMMAND` is restored to `starship_precmd`, the PS1 works as expected
- `starship_precmd() { :; }`
- the PS1 does not change anymore, we "broke" starship function
- `eval "$(cat ./src/init/starship.bash)"`
- the PS1 works as expected

</details>

 ---

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [x] I have tested using **Windows** (WSL)

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

I have 2 questions:
- I haven't change any Rust code, and my change does not require any configuration. Do I need to do something more?
- I don't have MacOS, could someone help in testing this please?

### CODE REVIEW

**Comment by NeodymiumFerBore** [comment]
**Problem**: if starship was init as part of bashrc (or any user profile script), vscode will define function `__vsc_preexec_all`, and use it as `DEBUG` trap, re-using custom traps. If not, and if no custom `DEBUG` trap is defined in user profile scripts, vscode defines instead `__vsc_preexec_only`.

**Current behavior**: if we are in vscode and starship was NOT init in a user profile script, `DEBUG` trap is replaced by starship trap, so it breaks some vscode features when running init during the session.

**New behavior**: if we are in vscode and starship was NOT init in a user profile script, then `starship_preexec` is never executed as part of the `DEBUG` trap. So, instead of breaking vscode, we break starship. This is unwanted.

Converting this PR as draft, it needs more digging.

**Comment by davidkna** [comment]
@NeodymiumFerBore Now that #5735 is merged, debug traps are avoided on recent bash versions.

**Comment by NeodymiumFerBore** [comment]
@davidkna thank you for your comment, I wish I saw it before haha, integration with vscode is very tricky because of this debug trap.

I will rebase my branch on master and only treat the `PROMPT_COMMAND` issue then. Also, will move my code to avoid skipping the new `PS0` part.

**Comment by NeodymiumFerBore** [comment]
Update:

Rebased onto master (f66bfd9435f215867681a699428bd882d8c63ce7 as of now)

Edited my tests: sourcing a file is not the same as evaluating it, and traps are not handled the same way. Tests done with `eval` instead of `source` (documentation says to use eval).

With this patch, evaluating multiple times `starship init bash` in a vscode terminal with shell integration does not trigger an infinite loop. However:
- with bash < 4.4, when `starship init bash` is re-evaluated, vscode features relying on DEBUG trap no longer work (command status icon, status history next to the scrollbar, etc.).
- with bash >= 4.4, when evaluating <ins>for the first time</ins> `starship init bash` manually in a vscode terminal (outside of user profile scripts), the same vscode features no longer work. It was already (and is still) the case, so that's not a regression.

For bash >= 4.4 evaluating `starship init bash` in a user profile script, nothing breaks/crashes. The fix is not perfect, but at least it prevents the crash described in #5816 in all cases

**Comment by davidkna** [comment]
I can't reproduce the issue, would you mind sharing more detailed instructions for reproducing the breakage in vscode?
